source/MaterialXGenOgsXml/GlslFragmentGenerator.cpp

```diff
@@ -237,31 +305,78 @@ ShaderPtr GlslFragmentGenerator::generate(const string& name, ElementPtr element
+
             if (!outputSocket->getType()->isFloat3())
             {
                 string finalOutput = outputSocket->getVariable() + "_tmp";
-                emitLine(_syntax->getTypeName(outputSocket->getType()) + " " + finalOutput + " = " + outputValue, stage);
+                emitLine(_syntax->getTypeName(outputSocket->getType()) + " " + finalOutput + " = " + outputValue, pixelStage);
                 toVec3(outputSocket->getType(), finalOutput);
-                emitLine("return " + finalOutput, stage);
+                emitLine("return " + finalOutput, pixelStage);
             }
             else
             {
-                emitLine("return " + outputValue, stage);
+                emitLine("return " + outputValue, pixelStage);
             }
         }
     }
 
     // End function
-    emitScopeEnd(stage);
+    emitScopeEnd(pixelStage);
 
     // Replace all tokens with real identifier names
-    replaceTokens(_tokenSubstitutions, stage);
+    replaceTokens(_tokenSubstitutions, pixelStage);
+
+    // Now emit private uniform definitions to a special stage which is only
+    // consumed by the HLSL cross-compiler.
+    //
+    const VariableBlock& privateUniforms = pixelStage.getUniformBlock(HW::PRIVATE_UNIFORMS);
+    if (!privateUniforms.empty())
+    {
+        ShaderStage& privateUniformsStage = shader->getStage(Stage::PRIVATE_UNIFORMS);
+        for (size_t i = 0; i < privateUniforms.size(); ++i)
+        {
+            const std::string& originalName = privateUniforms[i]->getVariable();
+            const std::string textureName = OgsXmlGenerator::samplerToTextureName(originalName);
+            if (!textureName.empty())
+            {
+                // GLSL for OpenGL (which MaterialX generates) uses combined
+                // samplers where the texture is implicit in the sampler.
+                // HLSL shader model 5.0 required by DX11 uses separate samplers
+                // and textures.
+                // We use a naming convention compatible with both OGS and
+                // SPIRV-Cross where we derive sampler names from texture names
+                // with a prefix and a suffix.
+                // Unfortunately, the cross-compiler toolchain converts GLSL
+                // samplers to HLSL textures preserving the original names and
+                // generates HLSL samplers with derived names. For this reason,
+                // we rename GLSL samplers with macros here to follow the
+                // texture naming convention which preserves the correct
+                // mapping.
+
+                emitLineBegin(privateUniformsStage);
+                emitString("#define ", privateUniformsStage);
+                emitString(originalName, privateUniformsStage);
+                emitString(" ", privateUniformsStage);
+                emitString(textureName, privateUniformsStage);
+                emitLineEnd(privateUniformsStage, false);
```

Comment:
Tricky remapping of samplers and textures between GLSL and HLSL while remaining compatible with OGS.